README.md
# Rough (Swift)

[![Version](https://img.shields.io/cocoapods/v/Rough.svg?style=flat)](http://cocoapods.org/pods/Rough)
[![License](https://img.shields.io/cocoapods/l/Rough.svg?style=flat)](http://cocoapods.org/pods/Rough)
[![Platform](https://img.shields.io/cocoapods/p/Rough.svg?style=flat)](http://cocoapods.org/pods/Rough)

Rough lets you draw in a sketchy, hand-drawn-like, style. It is Swift clone of [Rough.js](http://roughjs.com/). The library defines primitives to draw lines, curves, arcs, polygons, circles, and ellipses.


## Requirements
Xcode 9, iOS 10

## Installation

Rough is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod 'Rough'
```

## Usage
![Rectangle](https://raw.githubusercontent.com/bakhtiyork/Rough/master/docs/img/ex1.png)
```swift
let canvas = roughView.canvas
canvas.rectangle(origin: CGPoint(x: 10, y: 10), width: 200, height: 100)
roughView.setNeedsDisplay()
```

### Lines and Ellipses
![Lines and Ellipses](https://raw.githubusercontent.com/bakhtiyork/Rough/master/docs/img/ex2.png)
```swift
canvas.circle(center: CGPoint(x: 80, y: 120), radius: 25)
canvas.ellipse(center: CGPoint(x: 300, y: 100), width: 150, height: 80)
canvas.line(from: CGPoint(x: 80, y: 120), to: CGPoint(x: 300, y: 100))
```

### Filling
![Filling](https://raw.githubusercontent.com/bakhtiyork/Rough/master/docs/img/ex3.png)
```swift
canvas.circle(center: CGPoint(x: 50, y: 50), radius: 40) { options in
    options.fill = UIColor.red
} // fill with red hachure

canvas.rectangle(origin: CGPoint(x: 120, y: 15), width: 80, height: 80) { options in
    options.fill = UIColor.red
}

canvas.circle(center: CGPoint(x: 50, y: 150), radius: 40) {
    options in
    options.fill = UIColor(red: 10/255.0, green: 150/255.0, blue: 10/255.0, alpha: 1.0)
    options.fillWeight = 3 // thicker lines for hachure
}

canvas.rectangle(origin: CGPoint(x: 220, y: 15), width: 80, height: 80) { options in
    options.fill = UIColor.red
    options.hachureAngle = 60 // angle of hachure
    options.hachureGap = 8
}

canvas.rectangle(origin: CGPoint(x: 120, y: 105), width: 80, height: 80) { options in
    options.fill = UIColor(red: 1.0, green: 0, blue: 200/255.0, alpha: 0.2)
    options.fillStyle = .solid // solid fill
}
```


### Sketching style
![Sketching style](https://raw.githubusercontent.com/bakhtiyork/Rough/master/docs/img/ex4.png)
```swift
canvas.rectangle(origin: CGPoint(x: 15, y: 15), width: 80, height: 80) { options in
    options.roughness = 0.5
    options.fill = UIColor.red
}

canvas.rectangle(origin: CGPoint(x: 120, y: 15), width: 80, height: 80) { options in
    options.roughness = 2.8
    options.fill = UIColor.blue
}

canvas.rectangle(origin: CGPoint(x: 220, y: 15), width: 80, height: 80) { options in
    options.bowing = 6
    options.stroke = UIColor.green
    options.strokeWidth = 3
}
```

### SVG Path
*TODO*

## Example

To run the example project, clone the repo, and run `pod install` from the Example directory first.

<img alt="Screenshot1" src="https://raw.githubusercontent.com/bakhtiyork/Rough/master/docs/img/screenshot1.png" width="320"> <img alt="Screenshot2" src="https://raw.githubusercontent.com/bakhtiyork/Rough/master/docs/img/screenshot2.png" width="320">


## Credits

Credits to [Rough.js](http://roughjs.com/) by [Preet](https://github.com/pshihn)


## License

Rough is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Changelog

## 0.1.2
Increasing minimal iOS platform version to 10.0

## 0.1.1
Example app - using navigator controller to index samples

## 0.1.0
Initial release


Example/Tests/Tests.swift
import Rough
import XCTest

class Tests: XCTestCase {
    func testExample() {
    }

//    func testPerformanceExample() {
//        // This is an example of a performance test case.
//        self.measure() {
//            // Put the code you want to measure the time of here.
//        }
//    }
}


Rough/Classes/Rough.swift
//
//  Rough.swift
//  Rough
//
//  Copyright (c) 2018. MIT License
//  https://github.com/bakhtiyork/Rough
//

public typealias BezierPath = UIBezierPath

extension CGFloat {
    public var float: Float {
        return Float(self)
    }
}

extension FloatingPoint {
    var degreesToRadians: Self { return self * .pi / 180 }
    var radiansToDegrees: Self { return self * 180 / .pi }
}

extension String {
    mutating func removingRegexMatches(pattern: String, replaceWith: String = "") {
        do {
            let regex = try NSRegularExpression(pattern: pattern, options: NSRegularExpression.Options.caseInsensitive)
            let range = NSMakeRange(0, count)
            self = regex.stringByReplacingMatches(in: self, options: [], range: range, withTemplate: replaceWith)
        } catch {
            return
        }
    }
}

extension Float {
    public static var random: Float {
        return Float(arc4random()) / 0xFFFF_FFFF
    }
}

extension CGPoint {
    init(x: Float, y: Float) {
        self.init(x: CGFloat(x), y: CGFloat(y))
    }

    init(pair: (Float, Float)) {
        self.init(x: CGFloat(pair.0), y: CGFloat(pair.1))
    }
}

public enum DrawOp {
    case move(to: CGPoint)
    case bezierCurve(to: CGPoint, controlPoint1: CGPoint, controlPoint2: CGPoint)
    case quadCurve(to: CGPoint,
                   controlPoint: CGPoint)
    case line(to: CGPoint)
}

public enum Shape {
    case none
    case path(path: BezierPath, options: Options)
    case fill(path: BezierPath, options: Options)
    case hachure(path: BezierPath, options: Options)

    static func toBezierPath(ops: [DrawOp], closed: Bool) -> BezierPath {
        let bezierPath = BezierPath()
        for op in ops {
            switch op {
            case let .move(to: point):
                bezierPath.move(to: point)
            case let .bezierCurve(to: origin, controlPoint1: point1, controlPoint2: point2):
                bezierPath.addCurve(to: origin, controlPoint1: point1, controlPoint2: point2)
            case let .quadCurve(to: origin, controlPoint: point):
                bezierPath.addQuadCurve(to: origin, controlPoint: point)
            case let .line(to: point):
                bezierPath.addLine(to: point)
            }
        }
        if closed {
            bezierPath.close()
        }
        return bezierPath
    }
}

public enum FillType {
    case hachure, solid
}

@objc public class Options: NSObject {
    override init() {}

    public var maxRandomnessOffset: Float = 2.0
    public var bowing: Float = 1.0
    public var roughness: Float = 1.0
    public var stroke = UIColor.black
    public var strokeWidth: Float = 1.0
    public var curveTightness: Float = 0.0
    public var curveStepCount = 9
    public var fill = UIColor.clear
    public var fillStyle = FillType.hachure
    public var fillWeight: Float = 1.0
    public var hachureAngle: Float = -41.0
    public var hachureGap: Float = -1.0
    public var simplification: Bool = false

    public static func copy(options: Options) -> Options {
        let copy = Options()
        copy.maxRandomnessOffset = options.maxRandomnessOffset
        copy.bowing = options.bowing
        copy.roughness = options.roughness
        copy.stroke = options.stroke
        copy.strokeWidth = options.strokeWidth
        copy.curveTightness = options.curveTightness
        copy.curveStepCount = options.curveStepCount
        copy.fill = options.fill
        copy.fillStyle = options.fillStyle
        copy.fillWeight = options.fillWeight
        copy.hachureAngle = options.hachureAngle
        copy.hachureGap = options.hachureGap
        copy.simplification = options.simplification
        return copy
    }
}


Rough/Classes/Hachure.swift
//
//  Hachure.swift
//  Rough
//
//  Copyright (c) 2018. MIT License
//  https://github.com/bakhtiyork/Rough
//

enum SegmentRelation {
    case intersects, separate, undefined
}

let MIN_LIMIT: Float = 0.0001
let MAX_LIMIT: Float = 0.9999

struct Segment {
    var px1: Float = 0
    var py1: Float = 0
    var px2: Float = 0
    var py2: Float = 0
    var xi: Float = MAXFLOAT
    var yi: Float = MAXFLOAT
    var a: Float = 0
    var b: Float = 0
    var c: Float = 0
    var undefinded: Bool = false

    init(px1: Float, py1: Float, px2: Float, py2: Float) {
        self.px1 = px1
        self.py1 = py1
        self.px2 = px2
        self.py2 = py2
        a = py2 - py1
        b = px1 - px2
        c = px2 * py1 - px1 * py2
        undefinded = (a == 0) && (b == 0) && (c == 0)
    }

    mutating func compare(other: Segment) -> SegmentRelation {
        if undefinded || other.undefinded {
            return .undefined
        }
        var grad1 = MAXFLOAT
        var grad2 = MAXFLOAT
        var int1: Float = 0
        var int2: Float = 0

        if abs(b) > MIN_LIMIT {
            grad1 = -a / b
            int1 = -c / b
        }
        if abs(other.b) > MIN_LIMIT {
            grad2 = -other.a / other.b
            int2 = -other.c / other.b
        }

        if grad1 == MAXFLOAT {
            if grad2 == MAXFLOAT {
                if (-c / a) != (-other.c / other.a) {
                    return .separate
                }
                if (py1 >= min(other.py1, other.py2)) && (py1 <= max(other.py1, other.py2)) {
                    xi = px1
                    yi = py1
                    return .intersects
                }
                if (py2 >= min(other.py1, other.py2)) && (py2 <= max(other.py1, other.py2)) {
                    xi = px2
                    yi = py2
                    return .intersects
                }
                return .separate
            }
            xi = px1
            yi = grad2 * xi + int2
            if ((py1 - yi) * (yi - py2) < -MIN_LIMIT) || ((other.py1 - yi) * (yi - other.py2) < -MIN_LIMIT) {
                return .separate
            }
            if abs(other.a) < MIN_LIMIT {
                if (other.px1 - xi) * (xi - other.px2) < -MIN_LIMIT {
                    return .separate
                }
                return .intersects
            }
            return .intersects
        }

        if grad2 == MAXFLOAT {
            xi = other.px1
            yi = grad1 * xi + int1
            if ((other.py1 - yi) * (yi - other.py2) < -MIN_LIMIT) || ((py1 - yi) * (yi - py2) < -MIN_LIMIT) {
                return .separate
            }
            if abs(a) < MIN_LIMIT {
                if (px1 - xi) * (xi - px2) < -MIN_LIMIT {
                    return .separate
                }
                return .intersects
            }
            return .intersects
        }

        if grad1 == grad2 {
            if int1 != int2 {
                return .separate
            }
            if (px1 >= min(other.px1, other.px2)) && (px1 <= max(other.py1, other.py2)) {
                xi = px1
                yi = py1
                return .intersects
            }
            if (px2 >= min(other.px1, other.px2)) && (px2 <= max(other.px1, other.px2)) {
                xi = px2
                yi = py2
                return .intersects
            }
            return .separate
        }

        xi = (int2 - int1) / (grad1 - grad2)
        yi = grad1 * xi + int1

        if ((px1 - xi) * (xi - px2) < -MIN_LIMIT) || ((other.px1 - xi) * (xi - other.px2) < -MIN_LIMIT) {
            return .separate
        }
        return .intersects
    }

    func getLength() -> Float {
        let dx = px2 - px1
        let dy = py2 - py1
        return sqrt(dx * dx + dy * dy)
    }
}

class HachureIterator: IteratorProtocol {
    let top: Float
    let bottom: Float
    let left: Float
    let right: Float
    let gap: Float
    let sinAngle: Float
    let tanAngle: Float
    var pos: Float = 0
    var deltaX: Float = 0
    var hGap: Float = 0
    var sLeft: Segment?
    var sRight: Segment?

    init(top: Float, bottom: Float, left: Float, right: Float, gap: Float, sinAngle: Float, cosAngle: Float, tanAngle: Float) {
        self.top = top
        self.bottom = bottom
        self.left = left
        self.right = right
        self.gap = gap
        self.sinAngle = sinAngle
        self.tanAngle = tanAngle

        if abs(sinAngle) < MIN_LIMIT {
            pos = left + gap
        } else if abs(sinAngle) > MAX_LIMIT {
            pos = top + gap
        } else {
            deltaX = (bottom - top) * abs(tanAngle)
            pos = left - abs(deltaX)
            hGap = abs(gap / cosAngle)
            sLeft = Segment(px1: left, py1: bottom, px2: left, py2: top)
            sRight = Segment(px1: right, py1: bottom, px2: right, py2: top)
        }
    }

    func next() -> [Float]? {
        if abs(sinAngle) < MIN_LIMIT {
            if pos < right {
                let line = [pos, top, pos, bottom]
                pos += gap
                return line
            }
        } else if abs(sinAngle) > MAX_LIMIT {
            if pos < bottom {
                let line = [left, pos, right, pos]
                pos += gap
                return line
            }
        } else {
            var xLower = pos - deltaX / 2.0
            var xUpper = pos + deltaX / 2.0
            var yLower = bottom
            var yUpper = top
            if pos < (right + deltaX) {
                while (xLower < left && xUpper < left) || (xLower > right && xUpper > right) {
                    pos += hGap
                    xLower = pos - deltaX / 2.0
                    xUpper = pos + deltaX / 2.0
                    if pos > (right + deltaX) {
                        return nil
                    }
                }
                var s = Segment(px1: xLower, py1: yLower, px2: xUpper, py2: yUpper)
                if sLeft != nil && s.compare(other: sLeft!) == .intersects {
                    xLower = s.xi
                    yLower = s.yi
                }
                if sRight != nil && s.compare(other: sRight!) == .intersects {
                    xUpper = s.xi
                    yUpper = s.yi
                }
                if tanAngle > 0 {
                    xLower = right - (xLower - left)
                    xUpper = right - (xUpper - left)
                }
                let line = [xLower, yLower, xUpper, yUpper]
                pos += hGap
                return line
            }
        }
        return nil
    }
}


Rough/Classes/RoughView.swift
//
//  RoughView.swift
//  Rough
//
//  Copyright (c) 2018. MIT License
//  https://github.com/bakhtiyork/Rough
//

@objc open class RoughView: UIView {
    @objc public var canvas = RoughCanvas()

    @objc public required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }

    @objc public override init(frame: CGRect) {
        super.init(frame: frame)
    }

    @objc open override func draw(_ rect: CGRect) {
        super.draw(rect)
        if let context = UIGraphicsGetCurrentContext() {
            context.saveGState()
            if isOpaque, let backgroundColor = self.backgroundColor {
                backgroundColor.setFill()
                context.fill(rect)
            }
            context.setLineJoin(.bevel)
            context.setLineCap(.round)
            canvas.draw(context: context, rect: rect)
            context.restoreGState()
        }
    }
}


Example/Rough/AppDelegate.swift
//
//  AppDelegate.swift
//  Rough
//
//  Copyright (c) 2018. MIT License
//  https://github.com/bakhtiyork/Rough
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



Rough/Classes/RoughCanvas.swift
//
//  RoughCanvas.swift
//  Rough
//
//  Copyright (c) 2018. MIT License
//  https://github.com/bakhtiyork/Rough
//

public typealias OptionsCallback = (Options) -> Void

@objc open class RoughCanvas: NSObject {
    public var canvasOptions = Options()
    public var renderer = Renderer()
    var paths = [Shape]()

    public override init() {}

    @objc open func draw(context _: CGContext, rect _: CGRect) {
        for shape in paths {
            switch shape {
            case let Shape.path(path: path, options: options):
                options.stroke.setStroke()
                path.lineWidth = CGFloat(options.strokeWidth)
                path.stroke()
            case let Shape.fill(path: path, options: options):
                options.fill.setFill()
                path.fill()
            case let Shape.hachure(path: path, options: options):
                options.fill.setStroke()
                path.lineWidth = CGFloat(options.fillWeight)
                path.stroke()
            case Shape.none:
                break
            }
        }
    }

    @objc public func clear() {
        paths.removeAll()
    }

    @objc public func line(from: CGPoint, to: CGPoint) {
        drawLine(from: from, to: to, options: canvasOptions)
    }

    @objc public func line(from: CGPoint, to: CGPoint, optionsCallback: OptionsCallback) {
        let customOptions = Options.copy(options: canvasOptions)
        optionsCallback(customOptions)
        drawLine(from: from, to: to, options: customOptions)
    }

    @objc public func rectangle(origin: CGPoint, width: Float, height: Float) {
        drawRectangle(origin: origin, width: width, height: height, options: canvasOptions)
    }

    @objc public func rectangle(origin: CGPoint, width: Float, height: Float, optionsCallback: OptionsCallback) {
        let customOptions = Options.copy(options: canvasOptions)
        optionsCallback(customOptions)
        drawRectangle(origin: origin, width: width, height: height, options: customOptions)
    }

    @objc public func ellipse(center: CGPoint, width: Float, height: Float) {
        drawEllipse(center: center, width: width, height: height, options: canvasOptions)
    }

    @objc public func ellipse(center: CGPoint, width: Float, height: Float, optionsCallback: OptionsCallback) {
        let customOptions = Options.copy(options: canvasOptions)
        optionsCallback(customOptions)
        drawEllipse(center: center, width: width, height: height, options: customOptions)
    }

    @objc public func circle(center: CGPoint, radius: Float) {
        drawEllipse(center: center, width: 2 * radius, height: 2 * radius, options: canvasOptions)
    }

    @objc public func circle(center: CGPoint, radius: Float, optionsCallback: OptionsCallback) {
        let customOptions = Options.copy(options: canvasOptions)
        optionsCallback(customOptions)
        drawEllipse(center: center, width: 2 * radius, height: 2 * radius, options: customOptions)
    }

    @objc public func linearPath(points: [CGPoint]) {
        drawPath(points: points, close: false, options: canvasOptions)
    }

    @objc public func linearPath(points: [CGPoint], optionsCallback: OptionsCallback) {
        let customOptions = Options.copy(options: canvasOptions)
        optionsCallback(customOptions)
        drawPath(points: points, close: false, options: customOptions)
    }

    @objc public func polygon(points: [CGPoint]) {
        drawPath(points: points, close: true, options: canvasOptions)
    }

    @objc public func polygon(points: [CGPoint], optionsCallback: OptionsCallback) {
        let customOptions = Options.copy(options: canvasOptions)
        optionsCallback(customOptions)
        drawPath(points: points, close: true, options: customOptions)
    }

    @objc public func arc(center: CGPoint, width: Float, height: Float, start: Float, stop: Float, closed: Bool) {
        drawArc(center: center, width: width, height: height, start: start, stop: stop, closed: closed, options: canvasOptions)
    }

    @objc public func curve(points: [CGPoint]) {
        drawCurve(points: points, options: canvasOptions)
    }

    @objc public func curve(points: [CGPoint], optionsCallback: OptionsCallback) {
        let customOptions = Options.copy(options: canvasOptions)
        optionsCallback(customOptions)
        drawCurve(points: points, options: customOptions)
    }

    @objc public func arc(center: CGPoint, width: Float, height: Float, start: Float, stop: Float, closed: Bool, optionsCallback: OptionsCallback) {
        let customOptions = Options.copy(options: canvasOptions)
        optionsCallback(customOptions)
        drawArc(center: center, width: width, height: height, start: start, stop: stop, closed: closed, options: customOptions)
    }

    func drawLine(from: CGPoint, to: CGPoint, options: Options) {
        let shape = renderer.line(from: from, to: to, options: options)
        paths.append(shape)
    }

    func drawRectangle(origin: CGPoint, width: Float, height: Float, options: Options) {
        let x = origin.x.float
        let y = origin.y.float
        var points = [CGPoint]()
        points.append(origin)
        points.append(CGPoint(x: x + width, y: y))
        points.append(CGPoint(x: x + width, y: y + height))
        points.append(CGPoint(x: x, y: y + height))

        drawPath(points: points, close: true, options: options)
    }

    func drawEllipse(center: CGPoint, width: Float, height: Float, options: Options) {
        let shape = renderer.ellipse(point: center, width: width, height: height, options: options)
        if options.fill != UIColor.clear {
            if options.fillStyle == .hachure {
                let fill = renderer.hachureFillEllipse(center: center, width: width - options.strokeWidth, height: height - options.strokeWidth, options: options)
                paths.append(fill)
            } else if case let .path(path: path, options: options) = shape {
                let fill = Shape.fill(path: path, options: options)
                paths.append(fill)
            }
        }
        paths.append(shape)
    }

    func drawPath(points: [CGPoint], close: Bool, options: Options) {
        let shape = renderer.linearPath(points: points, close: close, options: options)
        if close && options.fill != UIColor.clear {
            if options.fillStyle == .hachure {
                let fill = renderer.hachureFillShape(points: points, options: options)
                paths.append(fill)
            } else {
                let fill = renderer.solidFillShape(points: points, options: options)
                paths.append(fill)
            }
        }
        paths.append(shape)
    }

    func drawArc(center: CGPoint, width: Float, height: Float, start: Float, stop: Float, closed: Bool, options: Options) {
        let shape = renderer.arc(center: center, width: width - options.strokeWidth, height: height - options.strokeWidth, start: start, stop: stop, closed: closed, roughClosure: true, options: options)
        if options.fill != UIColor.clear {
            if options.fillStyle == .hachure {
                let fill = renderer.hachureFillArc(center: center, width: width, height: height, start: start, stop: stop, options: options)
                paths.append(fill)
            } else {
                let path = renderer.arc(center: center, width: width - options.strokeWidth, height: height - options.strokeWidth, start: start, stop: stop, closed: true, roughClosure: false, options: options)

                if case let .path(path: path, options: options) = path {
                    let fill = Shape.fill(path: path, options: options)
                    paths.append(fill)
                }
            }
        }
        paths.append(shape)
    }

    func drawCurve(points: [CGPoint], options: Options) {
        let shape = renderer.curve(points: points, options: options)
        paths.append(shape)
    }
}


Example/Rough/ViewController.swift
//
//  ViewController.swift
//  Rough
//
//  Copyright (c) 2018. MIT License
//  https://github.com/bakhtiyork/Rough
//

import Rough
import UIKit

class ViewController: UIViewController {
    @IBOutlet var roughView: RoughView?

    override func viewDidLoad() {
        super.viewDidLoad()

        if let canvas = roughView?.canvas {
            // rectangle
            canvas.rectangle(origin: CGPoint(x: 10, y: 10), width: 100, height: 100)
            canvas.rectangle(origin: CGPoint(x: 140, y: 10), width: 100, height: 100) { options in
                options.fill = UIColor(red: 1.0, green: 0, blue: 0, alpha: 0.2)
                options.fillStyle = .solid
                options.roughness = 2
            }
            canvas.rectangle(origin: CGPoint(x: 10, y: 130), width: 100, height: 100) { options in
                options.fill = UIColor.red
                options.stroke = UIColor.blue
                options.hachureAngle = 60
                options.hachureGap = 10
                options.fillWeight = 5
                options.strokeWidth = 5
            }

            // ellipse and circle
            canvas.ellipse(center: CGPoint(x: 350, y: 50), width: 150, height: 80)
            canvas.ellipse(center: CGPoint(x: 610, y: 50), width: 150, height: 80) { options in
                options.fill = UIColor.blue
            }
            canvas.circle(center: CGPoint(x: 480, y: 50), radius: 40) { options in
                options.stroke = UIColor.red
                options.strokeWidth = 2
                options.fill = UIColor(red: 0, green: 1.0, blue: 0, alpha: 0.3)
                options.fillStyle = .solid
            }

            // overlapping circles
            canvas.circle(center: CGPoint(x: 480, y: 150), radius: 40) { options in
                options.stroke = UIColor.red
                options.strokeWidth = 4
                options.fill = UIColor(red: 0, green: 1.0, blue: 0, alpha: 1)
                options.fillWeight = 4
                options.hachureGap = 6
            }
            canvas.circle(center: CGPoint(x: 530, y: 150), radius: 40) { options in
                options.stroke = UIColor.blue
                options.strokeWidth = 4
                options.fill = UIColor(red: 1.0, green: 1.0, blue: 0, alpha: 1)
                options.fillWeight = 4
                options.hachureGap = 6
            }

            // linearPath and polygon
            canvas.linearPath(points: [CGPoint(x: 690, y: 10), CGPoint(x: 790, y: 20), CGPoint(x: 750, y: 120), CGPoint(x: 690, y: 100)]) { options in
                options.roughness = 0.7
                options.stroke = UIColor.red
                options.strokeWidth = 4
            }
            canvas.polygon(points: [CGPoint(x: 690, y: 130), CGPoint(x: 790, y: 140), CGPoint(x: 750, y: 240), CGPoint(x: 690, y: 220)])
            canvas.polygon(points: [CGPoint(x: 690, y: 250), CGPoint(x: 790, y: 260), CGPoint(x: 750, y: 360), CGPoint(x: 690, y: 340)]) { options in
                options.stroke = UIColor.red
                options.strokeWidth = 4
                options.fill = UIColor(red: 0, green: 0, blue: 1.0, alpha: 0.2)
                options.fillStyle = .solid
            }
            canvas.polygon(points: [CGPoint(x: 690, y: 370), CGPoint(x: 790, y: 385), CGPoint(x: 750, y: 480), CGPoint(x: 690, y: 460)]) { options in
                options.stroke = UIColor.red
                options.fill = UIColor(red: 0, green: 0, blue: 1.0, alpha: 0.6)
                options.hachureAngle = 65
            }

            // arcs
            canvas.arc(center: CGPoint(x: 350, y: 200), width: 200, height: 180, start: Float.pi, stop: Float.pi * 1.6, closed: false)
            canvas.arc(center: CGPoint(x: 350, y: 300), width: 200, height: 180, start: Float.pi, stop: Float.pi * 1.6, closed: true)
            canvas.arc(center: CGPoint(x: 350, y: 300), width: 200, height: 180, start: 0, stop: Float.pi / 2, closed: true) { options in
                options.stroke = UIColor.red
                options.strokeWidth = 4
                options.fill = UIColor(red: 1, green: 1, blue: 0, alpha: 0.4)
                options.fillStyle = .solid
            }
            canvas.arc(center: CGPoint(x: 350, y: 300), width: 200, height: 180, start: Float.pi / 2, stop: Float.pi, closed: true) { options in
                options.stroke = UIColor.blue
                options.strokeWidth = 2
                options.fillWeight = 4
                options.fill = UIColor(red: 1, green: 0, blue: 1, alpha: 0.4)
            }

            // draw sine curve
            var points = [CGPoint]()
            for i in 0 ... 20 {
                let x = (400.0 / 20.0) * Double(i) + 10.0
                let xdeg = (Double.pi / 100.0) * x
                let y = round(sin(xdeg) * 90.0) + 500.0
                points.append(CGPoint(x: x, y: y))
            }
            canvas.curve(points: points) { options in
                options.roughness = 1.2
                options.stroke = UIColor.red
                options.strokeWidth = 3
            }

            roughView?.setNeedsDisplay()
        }
    }
}
